### DISCUSSION TITLE
Feature Proposal: Configurable FP16 intermediate computation precision for inference

### DISCUSSION THREAD

**Discussion by blime4** [ggml-org/llama.cpp#19505] [Ideas]
Hi maintainers,

  I've been experimenting with running intermediate activations in FP16
  (instead of the current F32) on the CUDA backend, similar to how vLLM
  defaults to FP16 via `--dtype float16`. I wanted to check if this direction
  would be of interest before putting together any PRs.

  ### Background

  Currently, all CUDA operators (rms_norm, softmax, scale, etc.) assert 
  `src0->type == GGML_TYPE_F32` for activation tensors. Even with quantized 
  weights, activations stay in F32 throughout the graph.

  The idea is straightforward: cast activations to F16 at graph entry, keep 
  them in F16 through the compute graph, and cast back to F32 only for final 
  logits — the same approach vLLM and TensorRT-LLM use by default.

  ### What I have so far

  I have a local branch where I've added F16 input/output support to the 
  CUDA operators in the activation path (norm, softmax, rope, scale, etc.). 
  The existing F32 code paths are untouched — F16 is purely additive.

  In my testing with [model/hardware details], I'm seeing [rough numbers] 
  improvement in throughput, mainly from reduced memory bandwidth pressure 
  on the bandwidth-bound ops.

  ### Questions for maintainers

  - Is this something the project would be interested in?
  - If so, would you prefer this as one large PR or broken into smaller 
    pieces (e.g., operator-level F16 support first, then the graph-level 
    plumbing)?
  - Any concerns about the approach I should be aware of?

  Happy to share more details or benchmarks if this seems worth pursuing. 
  Thanks for your time!

**Comment by blime4** [comment]
I've successfully validated this approach locally on **Qwen2, Qwen3, and Qwen3-MoE** models — correctness passes, with measurable performance gains. I plan to start by enabling this `--dtype float16`-style feature for the Qwen model family.

---

Here's the PR list I intend to submit:

**PR 1 — Standalone operator FP16 support** (Ready)
- norm.cu, softmax.cu, rope.cu, scale.cu, getrows.cu, mean.cu, sumrows.cu, top-k.cu, cumsum.cu, set-rows.cu,  reduce_rows.cuh, common.cuh  
- [x] - https://github.com/ggml-org/llama.cpp/pull/19508 [close : need CPU only in the initial PR]

**PR 2 — Flash Attention FP16 accumulation paths** (Needs rebase)
- mma.cuh, fattn-common.cuh, fattn-mma-f16.cuh

**PR 3 — Mat-mul family FP16** (Needs rebase)
- mmvf.cu, mmvq.cu, mmf.cuh, mmf.cu, mmq.cuh

**PR 4 — Top-level dispatch + --compute-type option** (Planned)
- ggml-cuda.cu dispatch routing + user-facing CLI option

**Comment by JohannesGaessler** [comment]
I think a default compute type of FP16/BF16 instead of FP32 would be a useful feature but as is detailed in the contributing guidelines:

>When adding support for a new model or feature, focus on CPU support only in the initial PR unless you have a good reason not to. Add support for other backends like CUDA in follow-up PRs

This should IMO be done as something like a global setting for the ggml context in which the compute tensors are created and implemented on the CPU backend first so that there is a fallback.

>In my testing with [model/hardware details], I'm seeing [rough numbers]
improvement in throughput, mainly from reduced memory bandwidth pressure
on the bandwidth-bound ops.

Let me remind you of something else in the contributing guidelines:

>This project does not accept pull requests that are fully or predominantly AI-generated. AI tools may be utilized solely in an assistive capacity.

Technically copypasting language model outputs into the Discussions tab is not prohibited but I would consider this an oversight on our part.

**Comment by blime4** [comment]
@JohannesGaessler Thanks for the feedback! I'll close #19508  and submit a CPU-only PR for the scale and softcap FP16 support first.  I've tested it on an NVIDIA A40 and the tests pass. After the CPU PR is merged, I'll submit a follow-up PR adding the CUDA implementation.  Does this approach work? I will follow the contributing guidelines more in future Discussions .

**Comment by blime4** [comment]
@JohannesGaessler 
> This should IMO be done as something like a global setting for the ggml context in which the compute tensors are created and implemented on the CPU backend first so that there is a fallback.

I agree — a global compute_type setting on the ggml context is the right approach. In my local experiments with Qwen2, Qwen3, and Qwen3-MoE on Nvidia A40, I was doing the FP16 cast directly in the model graph builder (cast to F16 at graph entry, run the compute graph in F16, cast back to F32 before logits), which works but is clearly model-specific and not the right abstraction.

For example, in `src/models/qwen3.cpp`, my changes are:
```cpp
    inpL = build_inp_embd(model.tok_embd);

    // For fp16 intermediate computation precision
    ggml_type compute_type = cparams.use_fp16 ? GGML_TYPE_F16 : GGML_TYPE_F32;
    if (compute_type != GGML_TYPE_F32) {
        inpL = ggml_cast(ctx0, inpL, compute_type);
    }
```

```cpp
    // lm_head
    cur = build_lora_mm(model.output, cur);

    cb(cur, "result_output", -1);

    // For fp16 intermediate computation precision
    if (compute_type != GGML_TYPE_F32) {
        cur = ggml_cast(ctx0, cur, GGML_TYPE_F32);
        cb(cur, "result_output_f32", -1);
    }

    res->t_logits = cur;
```

**Comment by bssrdf** [comment]
May I suggest adding BF16 type in addition to FP16? Some models (flux diffusion) require ranges FP16's is not adequate.  This can easily be  done by a template parameter. 

I can confirm simply keeping intermediate tensors at Fp16/BF16 gave a decent speed-up.

**Reply by blime4** [reply]
> May I suggest adding BF16 type in addition to FP16? Some models (flux diffusion) require ranges FP16's is not adequate. This can easily be done by a template parameter.

@bssrdf Thanks for the reminder, I am currently writing a draft to support "This should IMO be done as something like a global setting for the ggml context as @JohannesGaessler reminder", but my own implementation of the CPU and cuda operator only verifies the correctness of the Qwen series model intermediate computation precision using FP16 and FP32, I don't know if there will be any problems with BF16, I will reserve BF16 for future use in the enumeration, maybe people in the community who are more familiar with this part of the code can help develop it.

**Reply by blime4** [reply]
https://github.com/ggml-org/llama.cpp/pull/19536 : to support "This should IMO be done as something like a global setting for the ggml context"

**Reply by JohannesGaessler** [reply]
Be aware that I am nowhere near the only maintainer and that other maintainers may disagree what the design should be.

**Reply by blime4** [reply]
@JohannesGaessler I know. If anything needs to be modified, I'm happy to cooperate

**Comment by ggerganov** [comment]
We'll implement this at some point and likely as @JohannesGaessler suggested it will be a `ggml_context` setting. It's something that has been experimented in the past and it can be done and it does provide some benefit in some scenarios - probably more relevant with newer architectures involving many unary and binary ops. However:

@blime4 I just don't have the confidence that you have the capacity to implement this feature. Stepping in a complex project with your AI tools and trying to implement such a fundamental feature is very likely not going to work and will likely waste our time. You need to first demonstrate some understanding of the codebase with smaller contributions and meaningful interactions with the long-term maintainers.

My recommendation is to focus your efforts on something else for the time being.

**Reply by blime4** [reply]
@ggerganov @JohannesGaessler thanks for your time and recommendation.
